aten/src/ATen/native/cuda/jit_utils.cpp

```diff
@@ -614,6 +651,8 @@ std::string generate_code(
   env.s("compute_type", compute_type);
   env.s("functor", func);
   env.s("name", name);
+  env.s("traits_string", get_traits_string());
```

Comment:
We'd like to make this more modular to keep the size of the CUDA C string as small as possible, since it's used to generate an SHA1 hash when caching, and SHA1 hash generation is computationally expensive in the length of the string.

@ngimel and I propose the following:

- A string allowing casting to/from complex types should be separated from the rest of complex support, and included when type conversions are supported (like today)
- The strings supporting complex mathematics should be included in the CUDA C if and only if the compute type is a complex type

Reply:
Follow-up: would you measure the performance of:

- calling complex sinc for the first time ever (uncached first call requiring just-in-time compilation)
- calling complex sinc for the first time in a new program (cached first call)
- calling complex since once the kernel has been placed on a device (second call or later)

As a point of comparison, on the same machine would you measure how long a precompiled sinc call takes in PyTorch today? We can just use one tensor size for all these, like a small 32 element tensor that lets us focus on the CPU time.

Reply:
I have done the change in https://github.com/pytorch/pytorch/pull/71577#discussion_r789253277, the benchmark in https://github.com/pytorch/pytorch/pull/71577#discussion_r789255776 I will do it later.

Reply:
Are `traits_string` and `cmath_string` needed in all the cases? It looks like you are putting it in absolutely all kernels